clustermesh-apiserver/etcdinit/root.go

```diff
@@ -0,0 +1,122 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package etcdinit
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"os/exec"
+	"path"
+	"syscall"
+	"time"
+
+	"github.com/cilium/cilium/clustermesh-apiserver/etcdinit/defaults"
+	einit "github.com/cilium/cilium/pkg/etcd/init"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+
+	"github.com/spf13/cobra"
+	client "go.etcd.io/etcd/client/v3"
+)
+
+var (
+	log                     = logging.DefaultLogger.WithField(logfields.LogSubsys, "etcdinit")
+	etcdBinaryLocation      = "/usr/bin/etcd"
+	etcdDataDir             = ""
+	etcdClusterName         = ""
+	etcdInitialClusterToken = ""
+	debug                   = false
+)
+
+func NewCmd() *cobra.Command {
+	rootCmd := &cobra.Command{
+		Use:   "etcdinit",
+		Short: "Initialise an etcd data directory for use by the etcd sidecar of clustermesh-apiserver",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return InitEtcdLocal()
+		},
+	}
+	rootCmd.Flags().StringVar(&etcdDataDir, "data-dir", defaults.EtcdDataDir, "etcd data directory")
+	rootCmd.Flags().StringVar(&etcdClusterName, "cluster-name", defaults.EtcdClusterName, "etcd cluster name")
+	rootCmd.Flags().StringVar(&etcdInitialClusterToken, "initial-cluster-token", defaults.EtcdInitialClusterToken, "etcd initial cluster token")
+	rootCmd.Flags().BoolVar(&debug, "debug", false, "Debug log output")
+	return rootCmd
+}
+
+func InitEtcdLocal() error {
+	ctx, cancelFn := context.WithTimeout(context.Background(), time.Minute*5)
+	defer cancelFn()
+
+	if debug {
+		logging.SetLogLevelToDebug()
+		log.Debug("Debug logging enabled")
+	}
+
+	log := log.
+		WithField("etcdBinaryLocation", etcdBinaryLocation).
+		WithField("etcdDataDir", etcdDataDir).
+		WithField("etcdClusterName", etcdClusterName).
+		WithField("etcdInitialClusterToken", etcdInitialClusterToken)
+
+	// When the clustermesh-apiserver is launched, we create a new etcd. We don't support persistence of any kind
+	// here. So we just delete the contents of the data directory. It should be empty, but if the initialization failed
+	// we might re-run this as an init container.
+	log.Info("Deleting contents of data directory")
+	dir, err := os.ReadDir(etcdDataDir)
+	if err != nil {
+		return err
+	}
+	for _, d := range dir {
+		log.WithField("path", d.Name()).Debug("Removing file/directory in data dir")
+		os.RemoveAll(path.Join(etcdDataDir, d.Name()))
```

Comment:
no error handling here, I guess we might remove the files directory prior to the file depending on the order of the ReadDir iteration?

Reply:
I don't think that is possible. My understanding is that `os.ReadDir` lists things in the given directory, and doesn't recur. So we wouldn't see files in a directory, just the directory. But, yes, errors could occur (I/O errors from disk problems, permission issues, etc.) I'll add handling and exit out if this fails.

Reply:
I've added error handling.